src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,556 @@
+  return impl_->total_compressed_size();
+}
+
+// row-group metadata
+class RowGroupMetaData::RowGroupMetaDataImpl {
+ public:
+  explicit RowGroupMetaDataImpl(const format::RowGroup* row_group)
+      : row_group_(row_group) {}
+  ~RowGroupMetaDataImpl() {}
+
+  inline int num_columns() const { return row_group_->columns.size(); }
+
+  inline int64_t num_rows() const { return row_group_->num_rows; }
+
+  inline int64_t total_byte_size() const { return row_group_->total_byte_size; }
+
+  std::unique_ptr<ColumnMetaData> Column(int i) {
+    DCHECK(i < num_columns()) << "The file only has " << num_columns()
+                              << " columns, requested metadata for column: " << i;
+    return ColumnMetaData::Make(
+        reinterpret_cast<const uint8_t*>(&row_group_->columns[i]));
+  }
+
+ private:
+  const format::RowGroup* row_group_;
+};
+
+std::unique_ptr<RowGroupMetaData> RowGroupMetaData::Make(
+    const uint8_t* metadata) {
+  return std::unique_ptr<RowGroupMetaData>(new RowGroupMetaData(metadata));
+}
+
+RowGroupMetaData::RowGroupMetaData(const uint8_t* metadata)
+    : impl_{std::unique_ptr<RowGroupMetaDataImpl>(new RowGroupMetaDataImpl(
+          reinterpret_cast<const format::RowGroup*>(metadata)))} {}
+RowGroupMetaData::~RowGroupMetaData() {}
+
+int RowGroupMetaData::num_columns() const {
+  return impl_->num_columns();
+}
+
+int64_t RowGroupMetaData::num_rows() const {
+  return impl_->num_rows();
+}
+
+int64_t RowGroupMetaData::total_byte_size() const {
+  return impl_->total_byte_size();
+}
+
+std::unique_ptr<ColumnMetaData> RowGroupMetaData::Column(int i) const {
+  return impl_->Column(i);
+}
+
+// file metadata
+class FileMetaData::FileMetaDataImpl {
+ public:
+  friend FileMetaDataBuilder;
+  FileMetaDataImpl() {}
+ 
+  explicit FileMetaDataImpl(const uint8_t* metadata, uint32_t* metadata_len) {
+    metadata_.reset(new format::FileMetaData);
+    DeserializeThriftMsg(metadata, metadata_len, metadata_.get());
+    InitSchema();
+  }
+  ~FileMetaDataImpl() {}
+
+  inline int num_columns() const { return schema_.num_columns(); }
+  inline int64_t num_rows() const { return metadata_->num_rows; }
+  inline int num_row_groups() const { return metadata_->row_groups.size(); }
+  inline int32_t version() const { return metadata_->version; }
+  inline const std::string& created_by() const { return metadata_->created_by; }
+  inline int num_schema_elements() const { return metadata_->schema.size(); }
+
+  void WriteTo(OutputStream* dst) { SerializeThriftMsg(metadata_.get(), 1024, dst); }
+
+  std::unique_ptr<RowGroupMetaData> RowGroup(int i) {
+    DCHECK(i < num_row_groups())
+        << "The file only has " << num_row_groups()
+        << " row groups, requested metadata for row group: " << i;
+    return RowGroupMetaData::Make(
+        reinterpret_cast<const uint8_t*>(&metadata_->row_groups[i]));
+  }
+
+  const SchemaDescriptor* schema() const { return &schema_; }
+
+ protected:
```

Comment:
Move both of these members to `private` and no protected block (the `friend` in private gives access to these